text/0013-prepare-blockbuilder-and-core-runtime-apis-for-mbms.md

```diff
@@ -0,0 +1,126 @@
+3. New callback `System::PostInherents`: Can replace `on_initialize`/`on_finalize` where a hard deadline is required (complements `poll`). It is guaranteed to execute each block.
+
+These three features can be implemented when fulfilling these two requirements:
+1. The runtime can tell the block author to not include any transactions in the block.
+2. The runtime can execute logic right after all pallet-provided inherents have been applied.
+
+## Stakeholders
+
+- Substrate Maintainers: They have to implement this, including tests, audit and
+  maintenance burden.
+- Polkadot Runtime developers: They will have to adapt the runtime files to this breaking change.
+- Polkadot Parachain Teams: They also have to adapt to the breaking changes but then eventually have
+  multi-block migrations available.
+
+## Explanation
+
+
+### `Core::initialize_block`
+
+This runtime API function is changed from returning `()` to `ExtrinsicInclusionMode`:
+```rust
+enum ExtrinsicInclusionMode {
+  /// All extrinsics are allowed in this block.
+  AllExtrinsics,
+  /// Only inherents are allowed in this block.
+  OnlyInherents,
+}
+```
+
+A block author MUST respect the `ExtrinsicInclusionMode` that is returned by `initialize_block`. The runtime MAY reject blocks that violate this requirement. 
+
+It is RECOMMENDED that block authors keep transactions in their transaction pool (if applicable)
+for as long as `initialize_block` returns `OnlyInherents`. The assumption is that these transactions become valid once the runtime finishes the MBM.  
+Backwards compatibility with the current runtime API SHOULD be implemented by block authors to not mandate a lockstep update of the authoring software.  
+This could be achieved by checking the runtime API version and assuming that `initialize_block` does not have a return value when the version is lower than 7.
+
+### `BlockBuilder::last_inherent`
+
+A block author MUST always invoke `last_inherent` directly after applying all runtime-provided inherents. The runtime MAY reject blocks that violate this requirement.
+
+### Combined
+
+Coming back to the three main features and how they can be implemented with these runtime APIs changes:
+
+**1. Multi-Block-Migrations**: The runtime is being put into lock-down mode for the duration of the migration process by returning `OnlyInherents` from `initialize_block`. This ensures that no user provided transaction can interfere with the migration process. It is absolutely necessary to ensure this, since otherwise a transaction could call into un-migrated storage and violate storage invariants. The entry-point for the MBM logic is `last_inherent`. This is a good spot, because any data that is touched in inherents, is not MBM-migratable anyway. It could also be done before all other inherents or at the end of the block in `finalize_block`, but there is no downside from doing it in `last_inherent` and the other two features are in favour of this.
```

Comment:
> This is a good spot, because any data that is touched in inherents, is not MBM-migratable anyway

This is not entirely true. Things like the parachain consensus system could still be migrated by a MBM. This would require that the `provide_inherent` call returns an empty "inherent", aka doesn't attach any data to it.

Reply:
I dont quite get what you mean. What i wanted to express is, that the logic of an inherent could try to decode storage that should currently not be accessed since an MBM is ongoing.

Reply:
Sentence was removed.